File: app/src/main/java/com/chooloo/www/callmanager/ui/contacts/ContactsFragment.java
package com.chooloo.www.callmanager.ui.contacts;

import android.database.Cursor;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.loader.content.Loader;

import com.chooloo.www.callmanager.cursorloader.FavoritesAndContactsLoader;
import com.chooloo.www.callmanager.entity.Contact;
import com.chooloo.www.callmanager.ui.contact.ContactBottomDialogFragment;
import com.chooloo.www.callmanager.ui.cursor.CursorFragment;

import static android.Manifest.permission.READ_CONTACTS;

public class ContactsFragment extends CursorFragment<ContactsAdapter> implements ContactsMvpView {

    private static final String[] REQUIRED_PERMISSIONS = {READ_CONTACTS};
    private final static String ARG_PHONE_NUMBER = "phoneNumber";
    private final static String ARG_CONTACT_NAME = "contactName";

    private ContactsPresenter<ContactsMvpView> mPresenter;

    public static ContactsFragment newInstance() {
        return ContactsFragment.newInstance(null, null);
    }

    public static ContactsFragment newInstance(@Nullable String phoneNumber, @Nullable String contactNumber) {
        ContactsFragment fragment = new ContactsFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PHONE_NUMBER, phoneNumber);
        args.putString(ARG_CONTACT_NAME, contactNumber);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public ContactsAdapter onGetAdapter() {
        ContactsAdapter contactsAdapter = new ContactsAdapter(activity);
        contactsAdapter.setOnContactItemClick(contact -> mPresenter.onContactItemClick(contact));
        contactsAdapter.setOnContactItemLongClickListener(contact -> mPresenter.onContactItemLongClick(contact));
        return contactsAdapter;
    }

    @Override
    public String[] onGetPermissions() {
        return REQUIRED_PERMISSIONS;
    }

    @Override
    public Loader<Cursor> onGetLoader(Bundle args) {
        String contactName = args != null ? args.getString(ARG_CONTACT_NAME, null) : null;
        String phoneNumber = args != null ? args.getString(ARG_PHONE_NUMBER, null) : null;
        return new FavoritesAndContactsLoader(activity, phoneNumber, contactName);
    }

    @Override
    public void onSetup() {
        super.onSetup();

        mLoaderId = 0;

        mPresenter = new ContactsPresenter<>();
        mPresenter.attach(this);

        load();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detach();
    }

    @Override
    public void load(@Nullable String phoneNumber, @Nullable String contactName) {
        Bundle args = new Bundle();
        args.putString(ARG_PHONE_NUMBER, phoneNumber);
        args.putString(ARG_CONTACT_NAME, contactName);
        setArguments(args);
        load();
    }

    @Override
    public void openContact(Contact contact) {
        ContactBottomDialogFragment.newInstance(contact).show(activity.getSupportFragmentManager(), ContactBottomDialogFragment.TAG);
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/contacts/ContactsFragment.kt
package com.chooloo.www.callmanager.ui.contacts

import androidx.lifecycle.ViewModelProvider
import com.chooloo.www.callmanager.adapter.ContactsAdapter
import com.chooloo.www.callmanager.entity.Contact
import com.chooloo.www.callmanager.livedata.ContactsLiveData
import com.chooloo.www.callmanager.ui.contact.ContactBottomDialogFragment
import com.chooloo.www.callmanager.ui.list.ListFragment
import com.chooloo.www.callmanager.util.runWithPermissions
import com.chooloo.www.callmanager.viewmodel.data.DataViewModel
import com.chooloo.www.callmanager.viewmodel.data.DataViewModelFactory

class ContactsFragment : ListFragment<ContactsAdapter>(), ContactsMvpView {

    private lateinit var _presenter: ContactsMvpPresenter<ContactsMvpView>
    private lateinit var _contactsLiveData: ContactsLiveData

    companion object {
        fun newInstance(): ContactsFragment = ContactsFragment()
    }

    override fun onGetAdapter(): ContactsAdapter {
        return ContactsAdapter(_activity).apply {
            setOnContactItemClick { contact -> _presenter.onContactItemClick(contact) }
            setOnContactItemLongClickListener { contact -> _presenter.onContactItemLongClick(contact) }
        }
    }

    override fun onSetup() {
        super.onSetup()

        _presenter = ContactsPresenter()
        _presenter.attach(this)

        _contactsLiveData = ViewModelProvider(this, DataViewModelFactory(_activity)).get(DataViewModel::class.java).contacts
        showEmptyPage(false)
        showNoPermissions(false)
        observe()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _presenter.detach()
    }

    override fun observe() = runWithPermissions(ContactsLiveData.REQUIRED_PERMISSION, callback = {
        _contactsLiveData.observe(viewLifecycleOwner, { contacts -> adapter.updateContacts(contacts) })
    })

    override fun openContact(contact: Contact) {
        ContactBottomDialogFragment.newInstance(contact).show(_activity.supportFragmentManager, ContactBottomDialogFragment.TAG)
    }
}